bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/HadoopHashTableSource.java

```diff
@@ -0,0 +1,464 @@
+
+    long numKeys = 0;
+    // Flag indicating that this workitem is finished.
+    boolean isDone = false;
+    ImmutableBytesWritable currentRangeStartKey;
+    // Hash for the current range.
+    ImmutableBytesWritable currentHash;
+    RangeHash currentRangeHash;
+
+    public HashBasedReader(
+        HadoopHashTableSource source,
+        ImmutableBytesWritable startRowInclusive,
+        ImmutableBytesWritable stopRowExclusive,
+        TableHashReader reader) {
+      this.reader = reader;
+      this.source = source;
+      this.startRowInclusive = startRowInclusive;
+      this.stopRowExclusive = stopRowExclusive;
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      // NO CHECKED EXCEPTIONS HERE.
+      LOG.debug(
+          "Starting a new reader at key range ["
+              + immutableBytesToString(startRowInclusive)
+              + " ,"
+              + immutableBytesToString(stopRowExclusive)
+              + ").");
+      numKeys = 0;
+
+      if (readNextKey()) {
+        // Dataflow calls start, followed by getCurrent. HashBased reader needs to read on TableHash
+        // twice to return a RangeHash since it specifies both range-start and range-end.
+        advance();
+        return true;
+      }
+
+      isDone = true;
+      return false;
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      if (isDone) {
+        LOG.debug("Ending workitem at key " + immutableBytesToString(currentRangeStartKey) + " .");
+        return false;
+      }
+
+      ImmutableBytesWritable startKey = this.currentRangeStartKey;
+      ImmutableBytesWritable hash = this.currentHash;
+
+      if (!readNextKey()) {
+        this.currentRangeHash = RangeHash.of(startKey, stopRowExclusive, hash);
+        // return true since we have lastBatchStartKey to emit. Set isDone=true to prevent reading
+        // from a potentially exhausted reader.
+        isDone = true;
+      } else {
+        this.currentRangeHash = RangeHash.of(startKey, reader.getCurrentKey(), hash);
+      }
+
+      return true;
+    }
+
+    // Returns true if a key can be read for this workitem.
+    private boolean readNextKey() throws IOException {
+      if (reader.next()) {
+        numKeys++;
+        this.currentRangeStartKey = reader.getCurrentKey();
+        if ( // StopRow is not set, everything is in bounds.
+        (stopRowExclusive.equals(HConstants.EMPTY_END_ROW)
+            || currentRangeStartKey.compareTo(stopRowExclusive) < 0)) { // currentKey < stopKey
+          // There is a key to read and the key is within the bounds of this workitem. Return true.
+          this.currentHash = reader.getCurrentHash();
+          return true;
+        } else {
+          // There is a key to read but its outside of the bounds of this workitem.
+          this.currentHash = null;
+          return false;
+        }
+      }
+
+      // Nothing left to read for this workitem.
```

Comment:
you have 2 end states: 1. you find the start of the next chunk. 2. you hit eof 

In case#1:
- reader.next() returns true 
- `numKeys++;` regardless if you hit the terminating row key.
- advance() makes a synthetic tail range

In case#2
- reader.next() returns false
- numKeys is unchanged
- advance() makes a synthetic tail range

This is inconsistent: you need to decide if you are counting the synthetic tail range or not